### ISSUE TITLE
Cannot find -ltinfo starting with release 1.38

### ISSUE THREAD

**Issue by robert-mijakovic** [hirooih/perl-trg#6]
I tried to build the package with Perl 5.32.1 however the build process doesn't work starting with the release 1.38.
I haven't changed the building procedure and have used the following commands:

```
$ tar xzf /home/apps/sources/p/Perl/extensions/Term-ReadLine-Gnu-1.40.tar.gz
$ perl Makefile.PL PREFIX=/home/apps/USE/easybuild/staging/2021.2/software/Perl/5.32.1-GCCcore-10.3.0

gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline -ltinfo
/home/apps/USE/easybuild/staging/2021.2/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: error: cannot find -ltinfo
collect2: error: ld returned 1 exit status
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Could not compile rlver.c.

system(): Inappropriate ioctl for device

If you have installed the GNU Readline Library (libreadline.{a,so} and
readline/readline.h, etc.) on directories for which your perl is not
configured to search (refer the value of `ccflags' and `libpath' in
the output of `perl -V'), specify the paths as follows;

        perl Makefile.PL --includedir=/yourdir/include --libdir=/yourdir/lib
or
        perl Makefile.PL --prefix=/yourdir

Note that the GNU Readline Library version 2.0 and earlier causes error
here.  Update it to version 2.1 and/or later.

Read INSTALL for more details.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!```


Exactly the same process with 1.37 works flawlessly.

**Comment by hirooih** [comment]
Thank you for your report.

Please provide me the information you asked for when you opened this issue.

https://github.com/hirooih/perl-trg/blob/master/.github/ISSUE_TEMPLATE/bug_report.md

Especially the following

> Output of
>   perl -V
>  perl Makefile.PL verbose

In your case provide me the result on `perl Makefile.PL verbose` on the release 1.38 in addition to above,

**Comment by robert-mijakovic** [comment]
Perl-Term::ReadLine::Gnu starting with 1.38 fails to find -ltinfo on our CentOS 8.3 machine with Perl 5.32.1 but 1.37 and earlier work just fine.

To Reproduce Steps to reproduce the behavior:
```
$ tar xzf /home/apps/sources/p/Perl/extensions/Term-ReadLine-Gnu-1.40.tar.gz
$ perl Makefile.PL PREFIX=/home/apps/USE/easybuild/staging/2021.2/software/Perl/5.32.1-GCCcore-10.3.0
```
I expected that it builds.

Output of
[perl -V](https://pastebin.com/KYBBe3te)
[perl Makefile.PL verbose](https://pastebin.com/7iYEwcsb) - perl-trg 1.37
[perl Makefile.PL verbose](https://pastebin.com/QZErATPa) - perl-trg 1.38

```
$ echo $TERM
xterm-256color
```

I'm using GNOME Terminal.
The GNU Readline Library (libreadline.a) was compiled with GCC 10.3.
Please note that everything works fine with TRG 1.37.

**Comment by hirooih** [comment]
```
[rmijakovic@login01 Term-ReadLine-Gnu-1.38]$ perl Makefile.PL verbose
Warning (mostly harmless): No library found for -ltermcap
...
```

TRG 1.38 fixed this issue; https://rt.cpan.org/Public/Bug/Display.html?id=134408.
Something wrong with `ExtUtils::Liblist->ext()` on your environment.

Could you apply the following patch and send me the output of `perl Makefile.PL verbose`?

```
--- a/Makefile.PL
+++ b/Makefile.PL
@@ -101,6 +101,7 @@ if ($Config{osname} eq 'os2') {
     # tinfo is for Debian. [Debian Bug Report #644423]
     my $PREFER_CURSES = $Config{osname} eq 'aix' || $Config{osname} eq 'hpux'
                        || $Config{osname} eq 'cygwin';
+    print "libpath:$Config{libpth}\n";
     my $TERMCAP_LIB = (! $PREFER_CURSES && &search_lib('-ltermcap'))
        || &search_lib('-ltinfo')
        || &search_lib('-lncurses')
@@ -199,7 +200,8 @@ exit(0);
 # ExtUtils::Liblist::ext() is not verbose by default any more and does much better job.
 sub search_lib {
     my ($lib) = @_;
-    my ($EXTRALIBS, $BSLOADLIBS, $LDLOADLIBS, $LD_RUN_PATH) = ExtUtils::Liblist->ext($lib);
+    my ($EXTRALIBS, $BSLOADLIBS, $LDLOADLIBS, $LD_RUN_PATH) = ExtUtils::Liblist->ext($lib, 1);
+    print "lib:$lib, EXTRALIBS:$EXTRALIBS, LDLOADLIB:$LDLOADLIBS\n";
     return $EXTRALIBS || $LDLOADLIBS || search_lib_sub($lib);
 }
```

**Comment by robert-mijakovic** [comment]
```
$ perl Makefile.PL verbose
``` 
[output](https://pastebin.com/hp9JhisN)

**Comment by hirooih** [comment]
Thanks.  We are getting closer.

Here is the comments in `Makefile.PL`.

```
    # Search libtermcap, libtinfo, libncurses, or libcurses in this order.
    # I emulate the behavior of the configure script for bash, and don't
    # know why AIX prefers curses.
```

1.37 (my original routine) checks only `lib*.a` and `lib*.so`.
But 1.38 uses `ExtUtils::Liblist->ext()` to support new OSX and it finds

```
'-ltinfo' found at /lib/../lib64/libtinfo.so.6.1
```

Your system should have `/lib64/libtinfo.so`, but does not. As result `ld` cannot find the library.

```
gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline -ltinfo
/usr/bin/ld.gold: error: cannot find -ltinfo
```

```
$ sudo ln -s /lib64/libtinfo.so.6.1  /lib64/libtinfo.so
```

should fix your system problem.

From `perl -V`

```
    libpth=/home/apps/USE/easybuild/staging/2021.2/software/DB/18.1.40-GCCcore-10.3.0/lib /home/apps/USE/easybuild/staging/2021.2/software/libreadline/8.1-GCCcore-10.3.0/lib /home/apps/USE/easybuild/staging/2021.2/software/ncurses/6.2-GCCcore-10.3.0/lib /home/apps/USE/easybuild/staging/2021.2/software/expat/2.2.9-GCCcore-10.3.0/lib /home/apps/USE/easybuild/staging/2021.2/software/zlib/1.2.11-GCCcore-10.3.0/lib /home/apps/USE/easybuild/staging/2021.2/software/binutils/2.36.1-GCCcore-10.3.0/lib /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib /usr/lib /usr/local/lib /home/apps/USE/easybuild/staging/2021.2/software/DB/18.1.40-GCCcore-10.3.0/lib/../lib64 /home/apps/USE/easybuild/staging/2021.2/software/libreadline/8.1-GCCcore-10.3.0/lib/../lib64 /home/apps/USE/easybuild/staging/2021.2/software/ncurses/6.2-GCCcore-10.3.0/lib/../lib64 /home/apps/USE/easybuild/staging/2021.2/software/expat/2.2.9-GCCcore-10.3.0/lib/../lib64 /home/apps/USE/easybuild/staging/2021.2/software/zlib/1.2.11-GCCcore-10.3.0/lib/../lib64 /home/apps/USE/easybuild/staging/2021.2/software/groff/1.22.4-GCCcore-10.3.0/lib/../lib64 /home/apps/USE/easybuild/staging/2021.2/software/binutils/2.36.1-GCCcore-10.3.0/lib/../lib64 /lib/../lib64 /usr/lib/../lib64 /home/apps/USE/easybuild/staging/2021.2/software/groff/1.22.4-GCCcore-10.3.0/lib /lib /lib64 /usr/lib64 /usr/local/lib64
```

This is interesting.  I don't know how Perl is configured like this.  If termcap is in the list before /usr/lib, it could be another fix.

**Comment by hirooih** [comment]
Another interesting behavior on your system.

from 1.37 result

```
$ perl Makefile.PL verbose
Found `/home/apps/USE/easybuild/staging/2021.2/software/ncurses/6.2-GCCcore-10.3.0/lib/libncurses.a'.
gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline -lncurses
```

How gcc (ld) could find ncurses library under a non-standard directory, `/home/apps/USE/easybuild/staging/2021.2/software/ncurses/6.2-GCCcore-10.3.0/lib/'?

**Comment by robert-mijakovic** [comment]
> Your system should have `/lib64/libtinfo.so`, but does not. As result `ld` cannot find the library.
> 
> ```
> gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline -ltinfo
> /usr/bin/ld.gold: error: cannot find -ltinfo
> ```
> 
> ```
> $ sudo ln -s /lib64/libtinfo.so.6.1  /lib64/libtinfo.so
> ```
> 
> should fix your system problem.

Yes, you are right. Our stock CentOS 8.3 installation doesn't have /lib64/libtinfo.so but rather:
```
$ ls -l /lib64/libtinfo.*
lrwxrwxrwx. 1 root root     15 Jan 16  2019 /lib64/libtinfo.so.5 -> libtinfo.so.5.9
-rwxr-xr-x. 1 root root 200000 Jan 16  2019 /lib64/libtinfo.so.5.9
lrwxrwxrwx. 1 root root     15 Jan 16  2019 /lib64/libtinfo.so.6 -> libtinfo.so.6.1
-rwxr-xr-x. 1 root root 208616 Jan 16  2019 /lib64/libtinfo.so.6.1
```

**Comment by robert-mijakovic** [comment]
Does TRG 1.37 depend on libtinfo? If it doesn't that explains why it didn't fail on my system.

**Comment by robert-mijakovic** [comment]
> Another interesting behavior on your system.
> 
> from 1.37 result
> 
> ```
> $ perl Makefile.PL verbose
> Found `/home/apps/USE/easybuild/staging/2021.2/software/ncurses/6.2-GCCcore-10.3.0/lib/libncurses.a'.
> gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline -lncurses
> ```
> 
> How gcc (ld) could find ncurses library under a non-standard directory, `/home/apps/USE/easybuild/staging/2021.2/software/ncurses/6.2-GCCcore-10.3.0/lib/'?

It is EasyBuild, a package manager that maintains the locations of dependency libraries.
Usually, ncurses built with the system compiler creates a symlink for libtinfo.so that points to libncurses.
```
$ ll /home/apps/USE/easybuild/staging/2021.1/software/ncurses/6.2/lib/libtinfo.so 
lrwxrwxrwx. 1 nobody nobody 78 Apr 14 21:19 /home/apps/USE/easybuild/staging/2021.1/software/ncurses/6.2/lib/libtinfo.so -> /home/apps/USE/easybuild/staging/2021.1/software/ncurses/6.2/lib/libncurses.so

```

**Comment by robert-mijakovic** [comment]
Ok, I can confirm that the issue is on the side of EasyBuild. 
It doesn't build ncurses with termlib and the version with non-system provided compiler doesn't create a symlink to it.
Sorry for bugging you with our bugs :)

**Comment by hirooih** [comment]
> Ok, I can confirm that the issue is on the side of EasyBuild.

I realized that this is the issue on my side.
`libreadline.so` knows a terminal library it needs. We don't have to (should not) specify the terminal library.

Could you try the following command lines, after `make Makefile.PL` fails?  (`-ltinfo` is removed.)

```
gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline
./rlver
```

If this works, I will make a fix of `Makefile.PL`.

**Comment by robert-mijakovic** [comment]
That is also true. If I'm not mistaken, the low-level terminal library is part of the libreadline symbols too. That is the reason the EasyBuild guys could create a symlink libtinfo that points to libncurses in some cases. I'll give it a try and let you know.

**Comment by robert-mijakovic** [comment]
Yes, the patch works.
```
--- Makefile.PL	2021-02-22 09:38:59.000000000 +0100
+++ Makefile.PL	2021-04-30 13:29:58.594223000 +0200
@@ -103,7 +103,6 @@
     my $PREFER_CURSES = $Config{osname} eq 'aix' || $Config{osname} eq 'hpux'
 			|| $Config{osname} eq 'cygwin';
     my $TERMCAP_LIB = (! $PREFER_CURSES && &search_lib('-ltermcap'))
-	|| &search_lib('-ltinfo')
 	|| &search_lib('-lncurses')
 	|| &search_lib('-lcurses');
```

**Comment by hirooih** [comment]
Thanks.  But I meant that execute the following commends after `perl Makefile.PL` fails.  (`perl Makefile.PL` creates `rlver.c`.)

```
gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline
./rlver
```

**Comment by robert-mijakovic** [comment]
```
$ gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtbegin.o: unknown program property type 0xc0010002 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtbegin.o: unknown program property type 0xc0010001 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtend.o: unknown program property type 0xc0010002 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtend.o: unknown program property type 0xc0010001 in .note.gnu.property section
$ ./rlver
8.1
```

**Comment by hirooih** [comment]
It works.  Give me some days to make the fix.
Thanks.

**Comment by hirooih** [comment]
Could you try this?  (Please rename to Makefile.PL)
[Makefile.txt](https://github.com/hirooih/perl-trg/files/6409156/Makefile.txt)

**Comment by robert-mijakovic** [comment]
Looks good to me.

```
$ perl Makefile.PL 
Warning (mostly harmless): No library found for -ltermcap
gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O rlver.c -o rlver    -fstack-protector-strong -L/usr/local/lib -lreadline -lncurses
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtbegin.o: unknown program property type 0xc0010002 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtbegin.o: unknown program property type 0xc0010001 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtend.o: unknown program property type 0xc0010002 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtend.o: unknown program property type 0xc0010001 in .note.gnu.property section
<<<It seems that you have the GNU Readline Library version 8.1.>>>
gcc  -D_REENTRANT -D_GNU_SOURCE -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2  -O -DTRG_READLINE_VERSION=0x0801  rlmalloc.c -o rlmalloc    -fstack-protector-strong -L/usr/local/lib -lreadline -lncurses
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtbegin.o: unknown program property type 0xc0010002 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtbegin.o: unknown program property type 0xc0010001 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtend.o: unknown program property type 0xc0010002 in .note.gnu.property section
/usr/bin/ld.gold: warning: /home/apps/USE/easybuild/staging/2021.2/software/GCCcore/10.3.0/lib/gcc/x86_64-pc-linux-gnu/10.3.0/crtend.o: unknown program property type 0xc0010001 in .note.gnu.property section
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for Term::ReadLine::Gnu
Writing MYMETA.yml and MYMETA.json
```

**Comment by hirooih** [comment]
I've released Term-ReadLine-Gnu-1.41.
Thank you for your report and helps.
